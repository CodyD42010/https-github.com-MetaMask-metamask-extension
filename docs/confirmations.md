# Adding New Confirmations

## Overview

Given the security focused nature of self-custody, confirmations and approvals form a pivotal aspect of the MetaMask extension.

Confirmations can be triggered by dApps and the UI itself, and are used to approve a variety of operations such as:

- Connecting to dApps
- Giving permissions to dApps
- Sending Eth
- Transfering tokens
- Signing data
- Interacting with Snaps
- Adding Ethereum networks

It is vital any new confirmations are implemented using best practices and consistent patterns, to avoid adding complexity to the code, and to minimise the maintenance cost of many alternate confirmations.

As we try to maintain a clean boundary between the UI and background page, the effort to implement a new confirmation can also be split accordingly.

## Background

### 1. Create Messenger

Ensure the controller or logic requiring a confirmation has access to a controller messenger.

Provide a `messenger` argument in the constructor if using a controller.

Ensure the allowed actions include at least the `ApprovalController:addRequest` action.

If the controller extends `BaseControllerV2`, the property `messagingSystem` is available to access the messenger passed to the base controller. Otherwise, ensure the provided messenger is assigned to a private property.

#### Example

```
this.someController = new SomeController({
    messenger: this.controllerMessenger.getRestricted({
      name: 'SomeController',
      allowedActions: [
        `${this.approvalController.name}:addRequest`,
        `${this.approvalController.name}:acceptRequest`,
        `${this.approvalController.name}:rejectRequest`,
      ],
    }),
    ...
});
```

### 2. Create Approval Request

Send an `addRequest` message to the `ApprovalController` to create an approval request.

This message returns a `Promise` which will resolve if the confirmation is approved, and reject if the confirmation is denied or cancelled.

Use an `async` function to send the message so the logic can `await` the confirmation and code execution can continue once approved. This enables the logic ran after approval to be kept in the same flow and therefore the logic to remain readable and encapsulated.

Ensure suitable error handling is in place to handle the confirmation being cancelled or denied and therefore the `Promise` being rejected.

The available message arguments are:

| Name | Description | Example Value |
| -- | -- | -- |
| opts.id | The ID of the approval request.<br>Assigned to a random value if not provided. | `"f81f5c8a-33bb-4f31-a4e2-52f8b94c393b"` |
| opts.origin | The origin of the request.<br>Either the dApp host or "metamask" if internal. | `"metamask.github.io"` |
| opts.type | An arbitrary string identifying the type of request. | `"eth_signTypedData"` |
| opts.requestData | Additional fixed data for the request.<br>Must be a JSON compatible object.| `{ transactionId: '123' }` |
| opts.requestState | Additional mutable data for the request.<br>Must be a JSON compatible object.<br>Can be updated using the `ApprovalController.updateRequestState` action. | `{ status: 'pending' }` |
| shouldShowRequest | A boolean indicating whether the popup should be displayed. | `true` |

#### Example

```
await this.messagingSystem.call(
  'ApprovalController:addRequest',
  {
    id,
    origin,
    type,
    requestData,
  },
  true,
);
```

### 3. Update Approval Request

If you wish to provide additional state to the confirmation while it is visible, send an `updateRequestState` message to the `ApprovalController`.

This requires you to have provided the `id` when creating the approval request, so it can be passed to the update message.

The available message arguments are:

| Name | Description | Example Value |
| -- | -- | -- |
| opts.id | The ID of the approval request to update. | `"f81f5c8a-33bb-4f31-a4e2-52f8b94c393b"` |
| opts.requestState | The updated mutable data for the request.<br>Must be a JSON compatible object. | `{ status: 'pending' }` |

#### Example

```
await this.messagingSystem.call(
  'ApprovalController:updateRequestState',
  {
    id,
    requestState: { counter },
  },
);
```

## Frontend

### 1. Create Template File

The `ConfirmationPage` component is already configured to display any approval requests generated by the `ApprovalController` and the associated `pendingApprovals` state.

In order to configure how the resulting confirmation is rendered, an **Approval Template** is required.

Create a new JavaScript file in `ui/pages/confirmation/templates` with the name matching the `type` used in the background approval request.

### 2. Update Approval Templates

Add your imported file to the `APPROVAL_TEMPLATES` constant in:
[ui/pages/confirmation/templates/index.js](../ui/pages/confirmation/templates/index.js)

### 3. Define Values

Inside the template file, define a `getValues` function that returns an object with the following properties:

| Name | Description | Example Value |
| -- | -- | -- |
| content | An array of objects defining the components to be rendered in the confirmation.<br>Processed by the [MetaMaskTemplateRenderer](../ui/components/app/metamask-template-renderer/metamask-template-renderer.js).  | See example below. |
| onSubmit | A callback to execute when the user approves the confirmation. | `actions.resolvePendingApproval(...)` |
| onCancel | A callback to execute when the user rejects the confirmation. | `actions.rejectPendingApproval(...)` |
| submitText | Text shown for the accept button. | `t('approveButtonText')` |
| cancelText | Text shown on the reject button. | `t('cancel')` |
| loadingText | Text shown while waiting for the onSubmit callback to complete. | `t('addingCustomNetwork')` |
| networkDisplay | A boolean indicating whether to show the current network at the top of the confirmation. | `true` |

#### Example

```
function getValues(pendingApproval, t, actions, _history) {
  return {
    content: [
      {
        element: 'Typography',
        key: 'title',
        children: 'Example',
        props: {
          variant: TypographyVariant.H3,
          align: 'center',
          fontWeight: 'bold',
          boxProps: {
            margin: [0, 0, 4],
          },
        },
      },
      ...
    ],
    cancelText: t('cancel'),
    submitText: t('approveButtonText'),
    loadingText: t('addingCustomNetwork'),
    onSubmit: () =>
      actions.resolvePendingApproval(
        pendingApproval.id,
        pendingApproval.requestData,
      ),
    onCancel: () =>
      actions.rejectPendingApproval(
        pendingApproval.id,
        ethErrors.provider.userRejectedRequest().serialize(),
      ),
    networkDisplay: true,
  };
}
```

### 4. Define Alerts

If any alerts are required in the confirmation, define the `getAlerts` function in the template file.

This needs to return an array of any required alerts, based on the current pending approval.

Each alert is an object with the following properties:

| Name | Description | Example Value |
| -- | -- | -- |
| id | A unique string to identify the alert.  | `"MISMATCHED_NETWORK_RPC"` |
| severity | The severity of the alert.<br>Use the constants from the design system. | `SEVERITIES.DANGER` |
| content | The component to be rendered inside the alert.<br>Uses the same format as the `content` returned from `getValues`.<br>The component can have nested components via the `children` property. | See example below. |

#### Example

```
function getAlerts(_pendingApproval) {
  return [
    {
      id: 'EXAMPLE_ALERT',
      severity: SEVERITIES.WARNING,
      content: {
        element: 'span',
        children: {
          element: 'MetaMaskTranslation',
          props: {
            translationKey: 'exampleMessage',
          },
        },
      },
    },
  ];
}
```

### 5. Export Functions

Ensure the `getValues` and `getAlerts` functions are exported from the template file.

#### Example

```
const example = {
  getAlerts,
  getValues,
};

export default example;
```

## Example Branch

See [this branch](https://github.com/MetaMask/metamask-extension/compare/develop...example/confirmation) as an example of the full code needed to add a confirmation.

The confirmation can be tested using the [E2E Test dApp](https://metamask.github.io/test-dapp/) and selecting `Request Permissions`.

## Glossary

### ApprovalController

The [ApprovalController](https://github.com/MetaMask/core/blob/main/packages/approval-controller/src/ApprovalController.ts) is a controller defined in the core repository which is responsible for creating and tracking approvals and confirmations in both the MetaMask extension and MetaMask mobile.

The `pendingApprovals` state used by the `ApprovalController` is not currently persisted, meaning any confirmations created by it will not persist after restarting the browser for example.

### ConfirmationPage

The [ConfirmationPage](../ui/pages/confirmation/confirmation.js) is a React component that aims to provide a generic confirmation window which can be configured using templates, each implementing a consistent interface.

This avoids the need for additional React components when creating confirmations, as additional templates with less logic, and less duplication, can be created instead.

## Screenshots

### Confirmation Window

[<img src="assets/confirmation.png" width="300">](assets/confirmation.png)
