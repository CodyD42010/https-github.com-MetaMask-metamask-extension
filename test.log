bigint: Failed to load bindings, pure JS will be used (try npm run rebuild?)
This version of µWS is not compatible with your Node.js build:

Error: Cannot find module './uws_darwin_arm64_108.node'
Falling back to a NodeJS implementation; performance may be degraded.


[driver] Called 'navigate' with arguments []
[driver] Called 'fill' with arguments ["#password","correct horse battery staple"]
[driver] Called 'press' with arguments ["#password",""]
[driver] Called 'openNewPage' with arguments ["https://metamask.github.io/snap-simple-keyring/0.4.0/"]
[driver] Called 'findElement' with arguments ["#connectButton"]
[driver] Called 'scrollToElement' with arguments [{"driver_":{"session_":{},"executor_":{"customCommands_":{},"log_":{"name_":"webdriver.http.Executor","level_":null,"parent_":{"name_":"webdriver.http","level_":null,"parent_":{"name_":"webdriver","l
[driver] Called 'waitUntilXWindowHandles' with arguments [3]
[driver] Called 'getAllWindowHandles' with arguments []
[driver] Called 'switchToWindowWithTitle' with arguments ["MetaMask Notification",["802eb8c9-6348-4f6f-96cf-035815ae78ac","9d558c38-99b6-4afe-890d-c28343bc0f49","ef58c194-e40f-4243-b6f9-7a34e9f747e8"]]
[driver] Called 'clickElement' with arguments [{"text":"Connect","tag":"button"}]
[driver] Called 'clickElementSafe' with arguments ["[data-testid=\"snap-install-scroll\"]",1000]
[driver] Called 'waitForSelector' with arguments [{"text":"Install"}]
[driver] Called 'clickElement' with arguments [{"text":"Install","tag":"button"}]
[driver] Called 'waitForSelector' with arguments [{"text":"OK"}]
[driver] Called 'clickElement' with arguments [{"text":"OK","tag":"button"}]
[driver] Called 'switchToWindowWithTitle' with arguments ["SSK - Simple Snap Keyring"]
[driver] Called 'waitForSelector' with arguments [{"text":"Connected","tag":"span"}]
[driver] Called 'switchToWindowWithTitle' with arguments ["SSK - Simple Snap Keyring"]
[driver] Called 'clickElement' with arguments [{"xpath":"//input[@id=\"use-sync-flow-toggle\"]/.."}]
[driver] Called 'clickElement' with arguments [{"text":"Create account","tag":"div"}]
[driver] Called 'clickElement' with arguments [{"text":"Create Account","tag":"button"}]
[driver] Called 'waitUntilXWindowHandles' with arguments [3]
[driver] Called 'getAllWindowHandles' with arguments []
[driver] Called 'switchToWindowWithTitle' with arguments ["MetaMask Notification",["802eb8c9-6348-4f6f-96cf-035815ae78ac","9d558c38-99b6-4afe-890d-c28343bc0f49","b9f26162-f19a-48d6-8bd2-e1957ffcfcc3"]]
[driver] Called 'clickElement' with arguments ["[data-testid=\"confirmation-submit-button\"]"]
[driver] Called 'clickElement' with arguments ["[data-testid=\"confirmation-submit-button\"]"]
[driver] Called 'switchToWindowWithTitle' with arguments ["SSK - Simple Snap Keyring"]
[driver] Called 'findElement' with arguments [{"text":"0x","tag":"p"}]
[driver] Called 'switchToWindowWithTitle' with arguments ["MetaMask"]
[driver] Called 'clickElement' with arguments ["[data-testid=\"account-menu-icon\"]"]
[driver] Called 'findElement' with arguments [{"css":".mm-tag","text":"Snaps (Beta)"}]
[driver] Called 'waitForElementNotPresent' with arguments [".mm-tag"]
[driver] Called 'openNewPage' with arguments ["http://127.0.0.1:8080"]
SNAPS/ signData called with approve true
[driver] Called 'switchToWindowWithTitle' with arguments ["E2E Test Dapp"]
[driver] Called 'clickElement' with arguments ["#personalSign"]
[driver] Called 'waitUntilXWindowHandles' with arguments [4]
[driver] Called 'getAllWindowHandles' with arguments []
[driver] Called 'switchToWindowWithTitle' with arguments ["MetaMask Notification",["802eb8c9-6348-4f6f-96cf-035815ae78ac","9d558c38-99b6-4afe-890d-c28343bc0f49","1580c9d0-7b11-4e47-99f0-3e4a07efc410","ce1954f5-e4f5-4fa8-b694-f2b375d959f4"]]
[driver] Called 'clickElement' with arguments [{"text":"Sign","tag":"button"}]
[driver] Called 'waitUntilXWindowHandles' with arguments [3]
NoSuchWindowError: Browsing context has been discarded
    at Object.throwDecodedError (/Users/owencraston/src/consensys/metamask-extension/node_modules/selenium-webdriver/lib/error.js:524:15)
    at parseHttpResponse (/Users/owencraston/src/consensys/metamask-extension/node_modules/selenium-webdriver/lib/http.js:601:13)
    at Executor.execute (/Users/owencraston/src/consensys/metamask-extension/node_modules/selenium-webdriver/lib/http.js:529:28)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at thenableWebDriverProxy.execute (/Users/owencraston/src/consensys/metamask-extension/node_modules/selenium-webdriver/lib/webdriver.js:745:17)
    at Driver.mockedEndpoint [as verboseReportOnFailure] (/Users/owencraston/src/consensys/metamask-extension/test/e2e/webdriver/driver.js:480:1)
    at withFixtures (/Users/owencraston/src/consensys/metamask-extension/test/e2e/helpers.js:157:7)
    at Context.<anonymous> (/Users/owencraston/src/consensys/metamask-extension/test/e2e/accounts/test-snap-accounts.spec.js:152:29) {
  remoteStacktrace: 'RemoteError@chrome://remote/content/shared/RemoteError.sys.mjs:8:8\n' +
    'WebDriverError@chrome://remote/content/shared/webdriver/Errors.sys.mjs:189:5\n' +
    'NoSuchWindowError@chrome://remote/content/shared/webdriver/Errors.sys.mjs:630:5\n' +
    'assert.that/<@chrome://remote/content/shared/webdriver/Assert.sys.mjs:485:13\n' +
    'assert.open@chrome://remote/content/shared/webdriver/Assert.sys.mjs:147:4\n' +
    'GeckoDriver.prototype.takeScreenshot@chrome://remote/content/marionette/driver.sys.mjs:2462:15\n' +
    'despatch@chrome://remote/content/marionette/server.sys.mjs:318:40\n' +
    'execute@chrome://remote/content/marionette/server.sys.mjs:289:16\n' +
    'onPacket/<@chrome://remote/content/marionette/server.sys.mjs:262:20\n' +
    'onPacket@chrome://remote/content/marionette/server.sys.mjs:263:9\n' +
    '_onJSONObjectReady/<@chrome://remote/content/marionette/transport.sys.mjs:494:20\n'
}
<testsuite name="Mocha Tests" tests="1" failures="0" errors="1" skipped="0" timestamp="Wed, 11 Oct 2023 22:49:46 GMT" time="43.19">
<testcase classname="Test Snap Account" name="can #personalSign (async flow approve)" time="43.188"><failure>waitUntilXWindowHandles timed out polling window handles
Error: waitUntilXWindowHandles timed out polling window handles
    at Driver.waitUntilXWindowHandles (test/e2e/webdriver/driver.js:408:3)
    at clickSignOnSignatureConfirmation (test/e2e/helpers.js:669:3)
    at withFixtures (test/e2e/accounts/test-snap-accounts.spec.js:241:9)
    at /Users/owencraston/src/consensys/metamask-extension/test/e2e/accounts/test-snap-accounts.spec.js:159:11
    at withFixtures (test/e2e/helpers.js:107:9)
    at Context.&#x3C;anonymous&#x3E; (test/e2e/accounts/test-snap-accounts.spec.js:152:29)</failure></testcase>
</testsuite>
