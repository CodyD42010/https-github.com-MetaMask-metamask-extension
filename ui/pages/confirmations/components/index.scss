/** Please import your files in alphabetical order **/
@import 'advanced-gas-inputs/index';
@import 'advanced-gas-controls/index';
@import 'cancel-speedup-popover/index';
@import 'confirm-page-container/index';
@import 'confirmation-warning-modal/index';
@import 'custom-nonce/index';
@import 'edit-gas-display/index';
@import 'edit-gas-fee-button/index';
@import 'edit-gas-fee-popover/index';
@import 'edit-gas-fee-popover/edit-gas-item/index';
@import 'edit-gas-fee-popover/network-statistics/index';
@import 'edit-gas-fee-popover/network-statistics/status-slider/index';
@import 'edit-gas-fee-popover/edit-gas-tooltip/index';
@import 'edit-gas-popover/index';
@import 'gas-details-item/index';
@import 'gas-timing/index';
@import 'multilayer-fee-message/index';
@import 'set-approval-for-all-warning/index';
@import 'signature-request/index';
@import 'signature-request-siwe/index';
@import 'signature-request-original/index';
@import 'signature-request-original/signature-request-original-warning/index';
@import 'transaction-decoding/index';
@import 'advanced-gas-fee-popover/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-gas-limit/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-inputs/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-inputs/base-fee-input/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-inputs/priority-fee-input/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-input-subtext/index';
@import 'advanced-gas-fee-popover/advanced-gas-fee-defaults/index';
@import 'approve-content-card/index';
@import 'transaction-alerts/transaction-alerts';
///: BEGIN:ONLY_INCLUDE_IF(conf-redesign)
@import 'confirm/header/header.scss';
///: END:ONLY_INCLUDE_IF
@import 'contract-details-modal/index';