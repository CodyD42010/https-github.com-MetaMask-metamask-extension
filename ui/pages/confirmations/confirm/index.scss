.confirm_wrapper {
  max-width: 719px;
  box-shadow: var(--shadow-size-lg) var(--color-shadow-default);
}