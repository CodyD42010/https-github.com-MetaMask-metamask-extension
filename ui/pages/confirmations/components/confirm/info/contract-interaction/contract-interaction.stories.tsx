import React from 'react';
import { Provider } from 'react-redux';
import { genUnapprovedContractInteractionConfirmation } from '../../../../../../../test/data/confirmations/contract-interaction';
import mockState from '../../../../../../../test/data/mock-state.json';
import configureStore from '../../../../../../store/store';
import ContractInteractionInfo from './contract-interaction';

const store = configureStore({
  metamask: { ...mockState.metamask },
  confirm: { currentConfirmation: genUnapprovedContractInteractionConfirmation() },
});

const Story = {
  title: 'Components/App/Confirm/info/ContractInteractionInfo',
  component: ContractInteractionInfo,
  decorators: [
    (story: () => any) => <Provider store={store}>{story()}</Provider>,
  ],
};

export default Story;

export const DefaultStory = () => <ContractInteractionInfo />;

DefaultStory.storyName = 'Default';
