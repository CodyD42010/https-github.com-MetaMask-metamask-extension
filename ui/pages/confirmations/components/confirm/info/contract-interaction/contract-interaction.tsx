import { TransactionMeta } from '@metamask/transaction-controller';
import React from 'react';
import { useSelector } from 'react-redux';
import { Box } from '../../../../../../components/component-library';
import {
  BackgroundColor,
  BorderRadius,
} from '../../../../../../helpers/constants/design-system';
import { currentConfirmationSelector } from '../../../../../../selectors';
import { SimulationDetails } from '../../../simulation-details';
import { TransactionDetails } from '../shared/transaction-details';

const ContractInteractionInfo: React.FC = () => {
  const currentConfirmation = useSelector(
    currentConfirmationSelector,
  ) as TransactionMeta;

  if (!currentConfirmation?.txParams) {
    return null;
  }

  return (
    <>
      <Box
        backgroundColor={BackgroundColor.backgroundDefault}
        borderRadius={BorderRadius.MD}
        marginBottom={4}
      >
        <SimulationDetails
          simulationData={currentConfirmation.simulationData}
          transactionId={currentConfirmation.id}
          isTransactionsRedesign
        />
      </Box>
      <Box
        backgroundColor={BackgroundColor.backgroundDefault}
        borderRadius={BorderRadius.MD}
        padding={2}
        marginBottom={4}
      >
        <TransactionDetails />
      </Box>
    </>
  );
};

export default ContractInteractionInfo;
