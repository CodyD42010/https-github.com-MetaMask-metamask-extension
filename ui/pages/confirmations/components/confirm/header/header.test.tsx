import { fireEvent, waitFor } from '@testing-library/react';
import React from 'react';
import { genUnapprovedContractInteractionConfirmation } from '../../../../../../test/data/confirmations/contract-interaction';
import { unapprovedPersonalSignMsg } from '../../../../../../test/data/confirmations/personal_sign';
import mockState from '../../../../../../test/data/mock-state.json';
import { renderWithProvider } from '../../../../../../test/jest';
import configureStore from '../../../../../store/store';
import Header from './header';

const render = (storeOverrides = {}) => {
  const store = configureStore({
    metamask: {
      ...mockState.metamask,
    },
    confirm: {
      currentConfirmation: {
        msgParams: {
          from: '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc',
        },
      },
    },
    ...storeOverrides,
  });

  return renderWithProvider(<Header />, store);
};

describe('Header', () => {
  it('should match snapshot with signature confirmation', () => {
    const { container } = render({
      confirm: { currentConfirmation: unapprovedPersonalSignMsg },
    });

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with transaction confirmation', () => {
    const { container } = render({
      confirm: {
        currentConfirmation: genUnapprovedContractInteractionConfirmation({
          address: '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc',
        }),
      },
    });

    expect(container).toMatchSnapshot();
  });

  it('contains network name and account name', () => {
    const { getByText } = render();
    expect(getByText('Test Account')).toBeInTheDocument();
    expect(getByText('Chain 5')).toBeInTheDocument();
  });
  it('contains account info icon', async () => {
    const { getByLabelText } = render();
    expect(getByLabelText('Account details')).toBeInTheDocument();
  });
  it('shows modal when account info icon is clicked', async () => {
    const { getByLabelText, queryByTestId } = render();
    expect(queryByTestId('account-details-modal')).not.toBeInTheDocument();
    const accountInfoIcon = getByLabelText('Account details');
    fireEvent.click(accountInfoIcon);
    await waitFor(() => {
      expect(queryByTestId('account-details-modal')).toBeInTheDocument();
    });
  });
});
