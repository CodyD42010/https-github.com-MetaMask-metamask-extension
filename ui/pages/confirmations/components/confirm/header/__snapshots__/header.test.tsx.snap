// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match snapshot with signature confirmation 1`] = `
<div>
  <div
    class="mm-box confirm_header__wrapper mm-box--display-flex mm-box--justify-content-space-between mm-box--align-items-center"
  >
    <div
      class="mm-box mm-box--padding-4 mm-box--display-flex mm-box--align-items-flex-start"
    >
      <div
        class="mm-box mm-box--margin-top-2 mm-box--display-flex"
      >
        <div
          class=""
        >
          <div
            class="identicon"
            style="height: 32px; width: 32px; border-radius: 16px;"
          >
            <div
              style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 32px; height: 32px; display: inline-block; background: rgb(3, 73, 94);"
            >
              <svg
                height="32"
                width="32"
                x="0"
                y="0"
              >
                <rect
                  fill="#FB184D"
                  height="32"
                  transform="translate(-0.010232866814926739 -2.467961361206475) rotate(274.3 16 16)"
                  width="32"
                  x="0"
                  y="0"
                />
                <rect
                  fill="#1888F2"
                  height="32"
                  transform="translate(0.7019171336752298 13.759175028660584) rotate(203.5 16 16)"
                  width="32"
                  x="0"
                  y="0"
                />
                <rect
                  fill="#FA8900"
                  height="32"
                  transform="translate(28.820219003567207 11.830135279564011) rotate(192.7 16 16)"
                  width="32"
                  x="0"
                  y="0"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-network confirm_header__avatar-network mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
        >
          C
        </div>
      </div>
      <div
        class="mm-box mm-box--margin-inline-start-4"
      >
        <p
          class="mm-box mm-text mm-text--body-md-medium mm-box--color-text-default"
        />
        <p
          class="mm-box mm-text mm-text--body-md mm-box--color-text-alternative"
        >
          Chain 5
        </p>
      </div>
    </div>
    <div
      class="mm-box mm-box--padding-4 mm-box--display-flex mm-box--align-items-flex-end"
    >
      <div
        class="mm-box mm-box--display-flex mm-box--justify-content-flex-end"
        style="align-self: flex-end;"
      >
        <div>
          <div
            aria-describedby="tippy-tooltip-1"
            class=""
            data-original-title="Account details"
            data-tooltipped=""
            style="display: inline;"
            tabindex="0"
          >
            <button
              aria-label="Account details"
              class="mm-box mm-button-icon mm-button-icon--size-md mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
            >
              <span
                class="mm-box mm-icon mm-icon--size-md mm-box--display-inline-block mm-box--color-inherit"
                style="mask-image: url('./images/icons/info.svg');"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Header should match snapshot with transaction confirmation 1`] = `
<div>
  <div
    class="mm-box confirm_header__wrapper mm-box--display-flex mm-box--justify-content-space-between mm-box--align-items-center"
  >
    <div
      class="mm-box mm-box--padding-4 mm-box--display-flex mm-box--align-items-flex-start"
    >
      <div
        class="mm-box mm-box--margin-top-2 mm-box--display-flex"
      >
        <div
          class=""
        >
          <div
            class="identicon"
            style="height: 32px; width: 32px; border-radius: 16px;"
          >
            <div
              style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 32px; height: 32px; display: inline-block; background: rgb(250, 58, 0);"
            >
              <svg
                height="32"
                width="32"
                x="0"
                y="0"
              >
                <rect
                  fill="#18CDF2"
                  height="32"
                  transform="translate(-1.04839350379394 -3.3042840694604987) rotate(328.9 16 16)"
                  width="32"
                  x="0"
                  y="0"
                />
                <rect
                  fill="#035E56"
                  height="32"
                  transform="translate(-18.298461708832043 10.5924618717486) rotate(176.2 16 16)"
                  width="32"
                  x="0"
                  y="0"
                />
                <rect
                  fill="#F26602"
                  height="32"
                  transform="translate(16.667842018223922 -14.205139722997082) rotate(468.9 16 16)"
                  width="32"
                  x="0"
                  y="0"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-network confirm_header__avatar-network mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
        >
          ?
        </div>
      </div>
      <div
        class="mm-box mm-box--margin-inline-start-4"
      >
        <p
          class="mm-box mm-text mm-text--body-md-medium mm-box--color-text-default"
        >
          Test Account
        </p>
        <p
          class="mm-box mm-text mm-text--body-md mm-box--color-text-alternative"
        />
      </div>
    </div>
    <div
      class="mm-box mm-box--padding-4 mm-box--display-flex mm-box--align-items-flex-end"
    >
      <div
        class="mm-box mm-box--display-flex mm-box--justify-content-flex-end"
        style="align-self: flex-end;"
      >
        <div>
          <div
            aria-describedby="tippy-tooltip-2"
            class=""
            data-original-title="Account details"
            data-tooltipped=""
            style="display: inline;"
            tabindex="0"
          >
            <button
              aria-label="Account details"
              class="mm-box mm-button-icon mm-button-icon--size-md mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
            >
              <span
                class="mm-box mm-icon mm-icon--size-md mm-box--display-inline-block mm-box--color-inherit"
                style="mask-image: url('./images/icons/info.svg');"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
